login_googleauth.8
.\" $Id: login_googleauth.8,v 1.1.1.1 2012/09/06 14:13:08 raxis Exp $
.\"
.\" Copyright (c) 2012 Chris Kruger <chris@krugerheavyindustries.com>
.\" All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 
.\"    - Redistributions of source code must retain the above copyright
.\"      notice, this list of conditions and the following disclaimer.
.\"    - Redistributions in binary form must reproduce the above
.\"      copyright notice, this list of conditions and the following
.\"      disclaimer in the documentation and/or other materials provided
.\"      with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
.\" FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
.\" COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
.\" BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
.\" LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
.\" CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
.\" ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\" 
.Dd $Mdocdate: September 6 2012 $
.Dt LOGIN_GOOGLEAUTH 8
.Os
.Sh NAME
.Nm login_googleauth
.Nd provides Google Authenticator authentication type
.Sh SYNOPSIS
.Nm login_googleauth
.Op Fl d
.Op Fl s Ar service
.Ar user
.Op Ar class
.Sh DESCRIPTION
The Google Authenticator provider supports the HMAC-based 
one-time password (HOTP) algorithm specified in RFC 4226 and 
the time-based on-time password (TOTP) specified in RFC 6238.
.Pp
The
.Nm
utility is called by
.Xr login 1 ,
.Xr su 1 ,
.Xr ftpd 8 ,
and others to authenticate the
.Ar user
with Google authentication.
.Pp
The
.Ar service
argument specifies which protocol to use with the
invoking program.
The allowed protocols are
.Em login ,
.Em challenge ,
and
.Em response .
The default protocol is
.Em login .
.Pp
The
.Ar user
argument is the login name of the user to be authenticated.
.Pp
The optional
.Ar class
argument is accepted for consistency with the other login scripts but
is not used.
.Pp
.Nm
will read the user's secret from the file
.Em user.secret ,
the user's last-use counter from
.Em user.counter
and the user's authenticator configuration from 
.Em user.conf
in the
.Em /var/db/googleauth
directory.
.Pp
If
.Ar user
does not have a secret file, the login is rejected.
If
.Ar user
does not have a last-use counter, a value of zero is used and
any counter is accepted during the first login.
.Pp
The one-time password provided by the user is decrypted using the
user's key.
After the decryption, the checksum embedded in the one-time password
is verified.
If the checksum is not valid, the login is rejected.
.Pp
If the checksum is valid, the uid embedded in the one-time password
is compared against the user's uid.
If the uid does not match, the login is rejected.
.Pp
If the uid matches, the use counter embedded in the one-time password
is compared to the last-use counter.
If the counter is less than or equal to the last-use counter, the
login is rejected.
This indicates a replay attack.
.Pp
If the counter is larger than the last-use counter, the counter
is stored as the new last-use counter, and the login is accepted.
.Pp
.Sh FILES
.Bl -tag -width /var/db/googleauth
.It Pa /var/db/googleauth
directory containing user entries for Googleauth
.El
.Sh SEE ALSO
.Xr login 1 ,
.Xr login.conf 5


login_googleauth.c
/*
 * bsd_auth login program supporting Google's two-factor authentication 
 *
 * Copyright (c) 2012 Chris Kruger <chris@krugerheavyindustries.com>
 * Portions Copyright (c) 2010 Daniel Hartmeier <daniel@benzedrine.cx>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *    - Redistributions of source code must retain the above copyright
 *      notice, this list of conditions and the following disclaimer.
 *    - Redistributions in binary form must reproduce the above
 *      copyright notice, this list of conditions and the following
 *      disclaimer in the documentation and/or other materials provided
 *      with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 */

#include <sys/param.h>
#include <sys/time.h>
#include <sys/resource.h>
#include <ctype.h>
#include <login_cap.h>
#include <pwd.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <unistd.h>

#include "common.h"
#include "googleauth.h"

#define	MODE_LOGIN	0
#define	MODE_CHALLENGE	1
#define	MODE_RESPONSE	2

#define	AUTH_OK		0
#define	AUTH_FAILED	-1

FILE *back = NULL;

#define DBPATH "/var/db/googleauth";

static int seperate_code_and_password(char *code, char *password, 
                                      const char *unified);

static int clean_string(const char *);
static int googleauth_login(const char *, const char *);

int main(int argc, char *argv[])
{
	int ch, ret, mode = MODE_LOGIN;

	char *username;
   char code[_PASSWORD_LEN];
   char password[1024];
	char response[1024];
   char *unified = NULL;
   char *wheel = NULL;
   char *class = NULL;

   int lastchance = 0;

   memset(password, 0, sizeof(password));
   memset(response, 0, sizeof(response));

	setpriority(PRIO_PROCESS, 0, 0);

	openlog(NULL, LOG_ODELAY, LOG_AUTH);

	while ((ch = getopt(argc, argv, "dv:s:")) != -1) {
		switch (ch) {
		case 'd':
			back = stdout;
			break;
		case 'v':
			syslog(LOG_INFO, "-v %s", optarg);
			break;
		case 's':
			if (!strcmp(optarg, "login"))
				mode = MODE_LOGIN;
			else if (!strcmp(optarg, "response"))
				mode = MODE_RESPONSE;
			else if (!strcmp(optarg, "challenge"))
				mode = MODE_CHALLENGE;
			else {
				syslog(LOG_ERR, "%s: invalid service", optarg);
				exit(EXIT_FAILURE);
			}
			break;
		default:
			syslog(LOG_ERR, "usage error1");
			exit(EXIT_FAILURE);
		}
	}
	argc -= optind;
	argv += optind;
	if (argc != 2 && argc != 1) {
		syslog(LOG_ERR, "usage error2");
		exit(EXIT_FAILURE);
	}
	username = argv[0];
	/* passed by sshd(8) for non-existing users */
	if (!strcmp(username, "NOUSER"))
		exit(EXIT_FAILURE);
	if (!clean_string(username)) {
		syslog(LOG_ERR, "clean_string username");
		exit(EXIT_FAILURE);
	}

	if (back == NULL && (back = fdopen(3, "r+")) == NULL) {
		syslog(LOG_ERR, "user %s: fdopen: %m", username);
		exit(EXIT_FAILURE);
	}

	switch (mode) {
		case MODE_LOGIN:
			if ((unified = getpass("Password:")) == NULL) {
				syslog(LOG_ERR, "user %s: getpass (Password)",
				    username);
				exit(EXIT_FAILURE);
			}
#if 0
         if ((password = getpass("Verification Code:")) == NULL) { 
            syslog(LOG_ERR, "user %s: getpass (Verification Code): %m",
                  username);
            exit(EXIT_FAILURE);
         }
         strncpy(code, password, _PASSWORD_LEN);
			if ((password = getpass("Password:")) == NULL) {
				syslog(LOG_ERR, "user %s: getpass (Password): %m",
				    username);
				exit(EXIT_FAILURE);
			}
#endif 
			break;
		case MODE_CHALLENGE:
			/* see login.conf(5) section CHALLENGES */
			fprintf(back, "%s\n", BI_SILENT);
         //fprintf(back, BI_VALUE " challenge %s\n", "Verification Code:");
         //fprintf(back, BI_CHALLENGE "\n");
         //fprintf(back, BI_FDPASS "\n");
         //fflush(back);
			exit(EXIT_SUCCESS);
			break;
		case MODE_RESPONSE: {
			/* see login.conf(5) section RESPONSES */
			/* this happens e.g. when called from sshd(8) */
			int count;
			mode = 0;
			count = -1;
			while (++count < sizeof(response) &&
			    read(fileno(back), &response[count], (size_t)1) ==
			    (ssize_t)1) {
				if (response[count] == '\0' && ++mode == 2)
					break;
				if (response[count] == '\0' && mode == 1) {
					unified = response + count + 1;
				}
			}
#if 0
         syslog(LOG_ERR, "got %s", unified);
#endif

			if (mode < 2) {
				syslog(LOG_ERR, "user %s: protocol error "
				    "on back channel", username);
				exit(EXIT_FAILURE);
			}
			break;
		}
	}

   if (seperate_code_and_password(code, password, unified)) {
      syslog(LOG_ERR, "user %s: incorrect format", username);
      exit(EXIT_FAILURE);
   }

	ret = googleauth_login(username, code); 

//   if (ret == AUTH_OK) 
//      ret = pwd_login(username, password, wheel, lastchance, class);

   memset(password, 0, sizeof(password));
   memset(response, 0, sizeof(response));
	if (unified != NULL){
      memset(unified, 0, strlen(unified));
	}
   if (ret != AUTH_OK) {
		syslog(LOG_INFO, "user %s: reject", username);
		fprintf(back, BI_REJECT "\n");
	}
	else { 
      syslog(LOG_INFO, "user %s: accepted", username);
		//This print statement returns the 'BI_AUTH' signal back to the bsd_auth
		//subsystem to proove the user account authenticated correctly. 
		//By putting this here, and commenting out the 'ret=pwd_login' statement
		//above, we shortcircuit the check to /etc/master.passwd for a user password
		//and rely ONLY on the GoogleAuth token for authentication. 
		fprintf(back, BI_AUTH "\n");
   }
	closelog();
	exit(EXIT_SUCCESS);
}

static int
seperate_code_and_password(char *code, char *password, const char *unified) 
{     
   // Verification are six digits starting with '0'..'9',
   // scratch codes are eight digits starting with '1'..'9'
   size_t len = strlen(unified);
   size_t span = strspn(unified, "0123456789");
   switch (span) {
      case 6:
         strncpy(code, unified, 6);
         strncpy(password, unified + 6, len - 6);
         return 0;
      case 8:
         strncpy(code, unified, 8);
         strncpy(password, unified + 8, len - 8);
         return 0;
   }
   return 1;
}

static int
clean_string(const char *s)
{
	while (*s) {
		if (!isalnum(*s) && *s != '-' && *s != '_')
			return (0);
		++s;
	}
	return (1);
}

static int
googleauth_login(const char *username, const char *password)
{
   uint8_t *secret = NULL;
   int      secretLen = 0;
   int      code = 0;
   char     *endptr;
   Config   config;

   if (read_cfg(username, &config)) {
		syslog(LOG_INFO, "user %s: error in user conf file", username);
      exit(EXIT_FAILURE);
   }

   long l = strtol(password, &endptr, 10);
   if (l < 0 || *endptr) {
      syslog(LOG_ERR, "user %s: failed to extract verification code", 
            username);
      return (AUTH_FAILED);
   }

   code = (int)l;

   if ((secret = get_shared_secret(username, &secretLen))) {
      int advance = 0;
      if (check_scratch_codes(username, code)) {
         if (config.is_totp) { 
            if (check_timebased_code(&config, secret, secretLen, code)) { 
#if 0 
               syslog(LOG_ERR, "user %s code %d: " 
                     "check_timebased_code != 0", 
                        username,
                        code); 
#else
               syslog(LOG_ERR, "user %s: " 
                     "check_timebased_code != 0", 
                        username);
#endif
               if (secret) {
                  memset(secret, 0, secretLen);
                  free(secret);
               }
               return (AUTH_FAILED);
            }
         } else { 
            int hotp_counter = get_hotp_counter(username);
            if (check_counterbased_code(&config, secret, secretLen, code,
                                        hotp_counter, &advance)) {
               syslog(LOG_ERR, "user %s: check_counterbased_code != 0",
                        username);
               if (secret) {
                  memset(secret, 0, secretLen);
                  free(secret);
               }
               return (AUTH_FAILED);
            }

            // If an hotp login attempt has been made, 
            // the counter must always be advanced by at least one
            if (advance) {
               if (advance_counter_by(username, hotp_counter, advance)) {
                  syslog(LOG_ERR, "user %s: failed to advance hotp counter",
                        username);
                  if (secret) { 
                     memset(secret, 0, secretLen);
                     free(secret);
                  }
                  return (AUTH_FAILED);
               }
            }
         }  
      }

      if (secret) {
         memset(secret, 0, secretLen);
         free(secret);
      }

      return (AUTH_OK);
   }

   return (AUTH_FAILED);
}
